plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodesTransportAction.java

```diff
@@ -126,22 +146,72 @@ MLStatsNodeResponse createMLStatsNodeResponse(MLStatsNodesRequest mlStatsNodesRe
             }
         }
 
+        CountDownLatch latch = new CountDownLatch(mlStats.getAllModels().length);
         Map<String, MLModelStats> modelStats = new HashMap<>();
         // return model level stats
         if (mlStatsInput.includeModelStats()) {
             for (String modelId : mlStats.getAllModels()) {
-                if (mlStatsInput.retrieveStatsForModel(modelId)) {
-                    Map<ActionName, MLActionStats> actionStatsMap = new HashMap<>();
-                    for (Map.Entry<ActionName, MLActionStats> entry : mlStats.getModelStats(modelId).entrySet()) {
-                        if (mlStatsInput.retrieveStatsForAction(entry.getKey())) {
-                            actionStatsMap.put(entry.getKey(), entry.getValue());
+                // Add action listener to retrieve model details
+                validateAccess(modelId, ActionListener.wrap(hasPermissionToShowStat -> {
+                    if (hasPermissionToShowStat) {
+                        if (mlStatsInput.retrieveStatsForModel(modelId)) {
+                            Map<ActionName, MLActionStats> actionStatsMap = new HashMap<>();
+                            for (Map.Entry<ActionName, MLActionStats> entry : mlStats.getModelStats(modelId).entrySet()) {
+                                if (mlStatsInput.retrieveStatsForAction(entry.getKey())) {
+                                    actionStatsMap.put(entry.getKey(), entry.getValue());
```

Comment:
can the users differentiate which model is the hidden model in the map of model_id and their values?

Reply:
user won't differentiate.